### ISSUE TITLE
Resolve problem with ThreadScopeContext in version 4.0.*

### ISSUE THREAD

**Issue by FinnKjar** [dadhi/DryIoc#151]
In my app I have multiple threads, and my container is created like this:

`_container = new Container(scopeContext: new ThreadScopeContext());`

I have two classes A and B. B takes a A in the ctor. I register B like this:

`_container.Register<B>();`

In each thread I have code like this:
```
                using (var scope = _container.OpenScope())
                {
                    var a = new A();
                    scope.UseInstance(a);
                    var b = _container.Resolve<B>();
                }
```
This works great using DryIoc 3.0.2, but using version 4.0.* an exception in thrown in some cases when doing the Resolve. It must be a sync problem. The exception is this:
```
Expecting the instance to be stored in singleton scope, but unable to find anything here.
Likely, you've called UseInstance from the scoped container, but resolving from another container or injecting into a singleton.
   at DryIoc.Throw.ThrowIfNull[T](T arg, Int32 error, Object arg0, Object arg1, Object arg2, Object arg3) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 10659
   at DryIoc.Container.InstanceFactory.GetInstanceFromScopeChainOrSingletons(IResolverContext r) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 1678
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, Object serviceKey, IfUnresolved ifUnresolved, Type requiredServiceType, Request preResolveParent, Object[] args) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 401
   at (Closure`1 , IResolverContext )
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 230
   at DryIoc.Resolver.Resolve[TService](IResolver resolver, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 5889
```

**Comment by dadhi** [comment]
Hi, try replace `scope.UseInstance(a)` with `scope.Use(a)`.

**Comment by FinnKjar** [comment]
Hi, thank you for a very quick response! I have tried to do as you suggested, but I still get an exception from time to time:
```
Unable to resolve A IsResolutionCall
  from Container with ambient ScopeContext DryIoc.ThreadScopeContext without Scope
Where no service registrations found
  and no dynamic registrations found in 0 of Rules.DynamicServiceProviders
  and nothing found in 0 of Rules.UnknownServiceResolvers
   at DryIoc.Throw.It(Int32 error, Object arg0, Object arg1, Object arg2, Object arg3) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 10696
   at DryIoc.Container.TryThrowUnableToResolve(Request request) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 936
   at DryIoc.Container.DryIoc.IContainer.ResolveFactory(Request request) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 916
   at DryIoc.Container.ResolveAndCacheFactoryDelegate(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 256
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 230
   at (IResolverContext )
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 230
   at DryIoc.Resolver.Resolve[TService](IResolver resolver, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.0.5\contentFiles\cs\net45\DryIoc\Container.cs:line 5889
```
And some times a wrong A is resolved, when resolving B. 
I have attached a simple console application, where I start up 20 threads doing some OpenScope(), Use/UseInstance and Resolve. 

[DryIocTest.zip](https://github.com/dadhi/DryIoc/files/3421432/DryIocTest.zip)

**Comment by dadhi** [comment]
Hmm, thanks - will check.

**Comment by dadhi** [comment]
May relate to #139 and #90

**Comment by dadhi** [comment]
Version 4.1.0-preview-01 is out. Could you check if it's working for you?

**Comment by FinnKjar** [comment]
Thank you! I will try the new version later today.

**Comment by FinnKjar** [comment]
I still have problems. Tried using the test application attached earlier. Get this exception:
`Unable to resolve Resolution root A IsResolutionCall
  from Container with ambient ScopeContext DryIoc.ThreadScopeContext without Scope
Where no service registrations found
  and no dynamic registrations found in 0 of Rules.DynamicServiceProviders
  and nothing found in 0 of Rules.UnknownServiceResolvers
   at DryIoc.Throw.It(Int32 error, Object arg0, Object arg1, Object arg2, Object arg3) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 11094
   at DryIoc.Container.TryThrowUnableToResolve(Request request) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 947
   at DryIoc.Container.DryIoc.IContainer.ResolveFactory(Request request) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 927
   at DryIoc.Container.ResolveAndCache(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 277
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 266
   at (IResolverContext )
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 243
   at DryIoc.Resolver.Resolve[TService](IResolver resolver, IfUnresolved ifUnresolved) in C:\Users\900143\.nuget\packages\dryioc\4.1.0-preview-01\contentFiles\cs\net45\DryIoc\Container.cs:line 6206
   at DryIocTest.Program.TestDryIoc() in C:\Users\900143\Source\repos\DryIocTest\Program.cs:line 77`

**Comment by FinnKjar** [comment]
Ups - hit the close button

**Comment by dadhi** [comment]
Thanks for checking, will check it myself.

**Comment by dadhi** [comment]
I have released v4.1.0-preview-3 with fix, please check it out.

**Comment by FinnKjar** [comment]
I works fine with preview-3. When is 4.1 to be released?
